File: Packages/Learning/Sources/Learning/EventLog.swift
import Foundation

public enum EventLogError: Error, Equatable, CustomStringConvertible {
    /// A word failed `EventLog.isLearnableWord` validation (empty, too long,
    /// contains whitespace/control characters, contains emoji/pictographs,
    /// or contains no letter at all). Pure-emoji and punctuation-only tokens
    /// are rejected by design: they are not vocabulary and would only inflate
    /// the personal store.
    case invalidContent(String)
    case ioError(String)

    public var description: String {
        switch self {
        case .invalidContent(let detail): return "Invalid event content: \(detail)"
        case .ioError(let detail): return "Event log I/O error: \(detail)"
        }
    }
}

/// Append-only, crash-safe learning event log.
///
/// Lives in the App Group container. The **keyboard extension is the only
/// writer** (append), the **containing app is the only compactor**
/// (read + truncate). The package takes a plain `URL` and performs no
/// coordination itself — see "Cross-process coordination" below.
///
/// ## On-disk format (schema version 1)
///
/// UTF-8 text, one record per `\n`-terminated line, tab-separated fields.
/// Field values escape `\` `\t` `\n` `\r` as `\\` `\t` `\n` `\r` (two-char
/// backslash sequences), so a real tab/newline never appears inside a field
/// and line/field framing is trivially recoverable.
///
/// ```
/// #gen<TAB><UUID>                                  ← generation header (first line)
/// 1<TAB><day><TAB>wc<TAB><word><TAB><prev|empty><TAB><is|en|un>
/// 1<TAB><day><TAB>sa<TAB><typed><TAB><accepted>
/// 1<TAB><day><TAB>cr<TAB><original><TAB><applied>
/// 1<TAB><day><TAB>wt<TAB><word>
/// 1<TAB><day><TAB>ts<TAB><keyChar><TAB><dx><TAB><dy>
/// ```
///
/// - The leading `1` is the per-line schema version; readers skip lines with
///   an unknown version, so future versions can mix records in one file.
/// - `<day>` is the coarse UTC `DayBucket` — the only temporal field
///   (privacy invariant: nothing finer than a day is ever written).
/// - The `#gen` header carries a random UUID minted whenever the file is
///   (re)created. Together with a byte offset it forms `ConsumedMarker`,
///   which lets the compactor prove "I already consumed bytes [0, offset)
///   *of this incarnation of the file*" across truncations and crashes.
///
/// ## Crash-safety guarantees
///
/// - **Appends** use a single `write(2)` on an `O_APPEND` descriptor with
///   the whole batch (one or more full lines) in one buffer. On APFS an
///   `O_APPEND` write of this size is not interleaved with other appends,
///   and a crash mid-write can only produce a torn suffix at EOF — never a
///   hole in the middle of the file.
/// - **Readers tolerate a torn final line**: only `\n`-terminated lines are
///   parsed; trailing unterminated bytes are ignored and excluded from the
///   returned `ConsumedMarker`.
/// - **Appends self-heal a torn tail**: before appending, the writer checks
///   the current last byte and inserts a `\n` if it isn't one, so the torn
///   fragment becomes an isolated garbage line (skipped by readers, counted
///   in `ReadResult.skippedLines`) instead of corrupting the next record.
///   The at-most-one-event loss window is a deliberate trade — no fsync per
///   keystroke.
/// - **Truncation** (`truncate(consumedUpTo:)`) rewrites the file atomically
///   (write-temp-then-rename) with a *new* generation UUID, keeping every
///   byte after the consumed offset. A compactor that crashes at any point
///   either re-reads only unconsumed events (generation mismatch resets the
///   offset to the header) or resumes from its stored marker — events are
///   never double-consumed and never lost. See
///   `PersonalModel.compactAndSave(applying:to:)` for the required ordering.
///
/// ## Cross-process coordination (App Group)
///
/// The App Group container is accessed by two processes. Callers on BOTH
/// sides must wrap every `append` / `read` / `truncate` in file coordination
/// (use `CoordinatedFileAccess`) so a truncation can never interleave with
/// an append. Two hard rules at the call boundary:
///
/// 1. Never hold a file descriptor for this URL across coordination blocks
///    (this type never does — every call opens and closes).
/// 2. Keep the extension's coordinated blocks short and synchronous: batch
///    events in memory and flush at word boundaries / lifecycle events, not
///    per keystroke. (Apple warns that long-held coordination in a suspended
///    extension can deadlock the coordinating app.)
public struct EventLog {
    public let url: URL

    /// Injectable day source (tests); defaults to the current UTC day bucket.
    public var dayProvider: () -> Int32

    public init(url: URL, dayProvider: @escaping () -> Int32 = { DayBucket.current() }) {
        self.url = url
        self.dayProvider = dayProvider
    }

    // MARK: - Consumed marker

    /// Position of the compactor's read frontier: "all complete lines of the
    /// file incarnation identified by `generation`, up to byte `offset`,
    /// have been merged into the personal model." Stored inside the
    /// `PersonalModel` file so consumption survives crashes exactly once.
    public struct ConsumedMarker: Codable, Equatable, Sendable {
        public let generation: UUID
        public let offset: UInt64

        public init(generation: UUID, offset: UInt64) {
            self.generation = generation
            self.offset = offset
        }

        /// Marker for a missing/empty log file.
        public static let none = ConsumedMarker(
            generation: UUID(uuid: (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)),
            offset: 0
        )
    }

    public struct ReadResult {
        public let events: [LoggedEvent]
        /// Marker positioned after the last complete line (feed this back
        /// into `read(after:)` / store it in the model as the consume-up-to
        /// point).
        public let endMarker: ConsumedMarker
        /// Complete lines that failed to parse (unknown schema version,
        /// healed torn fragments, hand-edited garbage). Skipped lines ARE
        /// covered by `endMarker` — they are consumed, just not applied.
        public let skippedLines: Int
    }

    // MARK: - Validation

    /// Whether a token is allowed into the learning log as a word.
    ///
    /// Rules: non-empty, at most `maxWordLength` characters, no whitespace or
    /// control characters (a "word" is a single token by definition), no
    /// extended-pictographic scalars (emoji are not vocabulary), and at least
    /// one letter (rejects digit-only and punctuation-only tokens).
    /// Non-ASCII letters — ð þ æ ö á é í ó ú ý and friends — are of course
    /// fine; this keyboard exists for them.
    public static let maxWordLength = 64

    public static func isLearnableWord(_ word: String) -> Bool {
        guard !word.isEmpty, word.count <= maxWordLength else { return false }
        var hasLetter = false
        for scalar in word.unicodeScalars {
            if CharacterSet.whitespacesAndNewlines.contains(scalar) { return false }
            if CharacterSet.controlCharacters.contains(scalar) { return false }
            // Emoji/pictograph rejection: `isEmoji` alone is true for ASCII
            // digits, so require a scalar beyond the basic ranges (0x203C is
            // the first emoji-capable scalar above ASCII/Latin).
            if scalar.properties.isEmojiPresentation
                || (scalar.properties.isEmoji && scalar.value >= 0x203C) {
                return false
            }
            if CharacterSet.letters.contains(scalar) { hasLetter = true }
        }
        return hasLetter
    }

    // MARK: - Append (keyboard extension side)

    /// Append a single event. See `append(contentsOf:)`.
    public func append(_ event: LearningEvent) throws {
        try append(contentsOf: [event])
    }

    /// Append a batch of events as one atomic `O_APPEND` write.
    ///
    /// Words are validated with `isLearnableWord`; an invalid primary word
    /// throws `EventLogError.invalidContent` (nothing is written). An invalid
    /// `previousWord` is downgraded to `nil` — the primary word is still
    /// worth learning even when its predecessor was an emoji or a number.
    ///
    /// PRIVACY: the caller must have verified the field context first — see
    /// `LearningPrivacy.assertLoggableFieldContext`.
    public func append(contentsOf events: [LearningEvent]) throws {
        guard !events.isEmpty else { return }
        let day = dayProvider()
        var buffer = ""
        for event in events {
            buffer += try Self.encodeLine(event, day: day)
        }
        try appendRaw(buffer)
    }

    // MARK: - Read (containing app / compactor side)

    /// Read all complete lines after `marker` (or after the generation
    /// header when `marker` is nil or from a different file incarnation).
    /// Tolerates a torn final line; never mutates the file.
    public func read(after marker: ConsumedMarker? = nil) throws -> ReadResult {
        guard FileManager.default.fileExists(atPath: url.path) else {
            return ReadResult(events: [], endMarker: .none, skippedLines: 0)
        }
        let bytes: [UInt8]
        do {
            bytes = [UInt8](try Data(contentsOf: url))
        } catch {
            throw EventLogError.ioError("read failed: \(error)")
        }
        let (generation, headerEnd) = Self.parseHeader(bytes)
        guard let headerEnd else {
            // Header line itself is torn (file created but the very first
            // write was interrupted before its newline). Nothing consumable.
            return ReadResult(events: [], endMarker: .none, skippedLines: 0)
        }

        var start = headerEnd
        if let marker, marker.generation == generation {
            let offset = Int(marker.offset)
            if offset >= headerEnd && offset <= bytes.count {
                start = offset
            }
        }

        var events: [LoggedEvent] = []
        var skipped = 0
        var cursor = start
        var consumedEnd = start
        while cursor < bytes.count {
            guard let newlineIndex = bytes[cursor...].firstIndex(of: 0x0A) else {
                break  // torn final line — leave it out of the marker
            }
            let line = String(decoding: bytes[cursor..<newlineIndex], as: UTF8.self)
            if let logged = Self.decodeLine(line) {
                events.append(logged)
            } else if !line.isEmpty {
                skipped += 1
            }
            cursor = newlineIndex + 1
            consumedEnd = cursor
        }
        return ReadResult(
            events: events,
            endMarker: ConsumedMarker(generation: generation, offset: UInt64(consumedEnd)),
            skippedLines: skipped
        )
    }

    // MARK: - Truncate (containing app / compactor side)

    /// Drop the consumed prefix `[header, marker.offset)`, preserving every
    /// byte a concurrent writer appended after the compactor's read. The
    /// file is rewritten atomically with a fresh generation UUID and the
    /// unconsumed tail; returns the marker for "nothing of the new file
    /// consumed yet".
    ///
    /// Safe orderings (all enforced by `PersonalModel.compactAndSave`):
    /// - Crash before: model already saved with the old-generation marker →
    ///   next compaction resumes after `marker.offset`, no double count.
    /// - Crash after: model still holds the old-generation marker, but the
    ///   file now has a new generation → the mismatch resets the read to the
    ///   header, and the file contains only unconsumed events. No loss, no
    ///   double count.
    ///
    /// If the file's generation no longer matches `marker.generation` (the
    /// log was already rotated) this is a no-op and returns `marker`
    /// unchanged. Must run inside the same coordinated-write block as the
    /// `read` that produced `marker` (see `CoordinatedFileAccess`), so no
    /// append can land between read and truncate unseen.
    @discardableResult
    public func truncate(consumedUpTo marker: ConsumedMarker) throws -> ConsumedMarker {
        guard FileManager.default.fileExists(atPath: url.path) else { return marker }
        let bytes: [UInt8]
        do {
            bytes = [UInt8](try Data(contentsOf: url))
        } catch {
            throw EventLogError.ioError("read for truncate failed: \(error)")
        }
        let (generation, headerEnd) = Self.parseHeader(bytes)
        guard headerEnd != nil, generation == marker.generation else { return marker }

        let tailStart = min(Int(marker.offset), bytes.count)
        let newGeneration = UUID()
        var newContents = Data(Self.headerLine(generation: newGeneration).utf8)
        let newOffset = UInt64(newContents.count)
        if tailStart < bytes.count {
            newContents.append(contentsOf: bytes[tailStart...])
        }
        do {
            try newContents.write(to: url, options: .atomic)
        } catch {
            throw EventLogError.ioError("truncate rewrite failed: \(error)")
        }
        return ConsumedMarker(generation: newGeneration, offset: newOffset)
    }

    // MARK: - Line codec

    static func headerLine(generation: UUID) -> String {
        "#gen\t\(generation.uuidString)\n"
    }

    /// Returns (generation, byte offset just past the header's newline).
    /// `headerEnd == nil` means the header line is torn/unreadable.
    /// A file that doesn't start with `#gen` is treated as generation
    /// `.none` with the data starting at offset 0.
    static func parseHeader(_ bytes: [UInt8]) -> (generation: UUID, headerEnd: Int?) {
        let prefix = Array("#gen\t".utf8)
        guard bytes.count >= prefix.count, Array(bytes[0..<prefix.count]) == prefix else {
            // Empty file, or a file that predates the header — treat all
            // bytes as data belonging to the "none" generation.
            return (ConsumedMarker.none.generation, 0)
        }
        guard let newlineIndex = bytes.firstIndex(of: 0x0A) else {
            return (ConsumedMarker.none.generation, nil)
        }
        let uuidString = String(decoding: bytes[prefix.count..<newlineIndex], as: UTF8.self)
        guard let uuid = UUID(uuidString: uuidString) else {
            return (ConsumedMarker.none.generation, newlineIndex + 1)
        }
        return (uuid, newlineIndex + 1)
    }

    static func encodeLine(_ event: LearningEvent, day: Int32) throws -> String {
        func validated(_ word: String, field: String) throws -> String {
            guard isLearnableWord(word) else {
                throw EventLogError.invalidContent("\(field) is not a learnable word")
            }
            return escape(word)
        }
        let fields: [String]
        switch event {
        case .wordCommitted(let word, let previousWord, let hint):
            let prev: String
            if let previousWord, isLearnableWord(previousWord) {
                prev = escape(previousWord)
            } else {
                prev = ""  // invalid predecessor downgrades to nil
            }
            fields = ["wc", try validated(word, field: "word"), prev, hint.rawValue]
        case .suggestionAccepted(let typed, let accepted):
            fields = ["sa", try validated(typed, field: "typed"), try validated(accepted, field: "accepted")]
        case .correctionReverted(let original, let applied):
            fields = ["cr", try validated(original, field: "original"), try validated(applied, field: "applied")]
        case .wordTapped(let word):
            fields = ["wt", try validated(word, field: "word")]
        case .touchSample(let keyChar, let dx, let dy):
            fields = [
                "ts",
                escape(String(keyChar)),
                String(format: "%.4f", dx),
                String(format: "%.4f", dy),
            ]
        }
        return "1\t\(day)\t" + fields.joined(separator: "\t") + "\n"
    }

    static func decodeLine(_ line: String) -> LoggedEvent? {
        let fields = line.split(separator: "\t", omittingEmptySubsequences: false).map(String.init)
        guard fields.count >= 3, fields[0] == "1", let day = Int32(fields[1]) else { return nil }
        switch fields[2] {
        case "wc":
            guard fields.count == 6, let hint = LanguageHint(rawValue: fields[5]) else { return nil }
            let word = unescape(fields[3])
            let prev = fields[4].isEmpty ? nil : unescape(fields[4])
            guard !word.isEmpty else { return nil }
            return LoggedEvent(day: day, event: .wordCommitted(word: word, previousWord: prev, languageHint: hint))
        case "sa":
            guard fields.count == 5 else { return nil }
            let typed = unescape(fields[3]), accepted = unescape(fields[4])
            guard !typed.isEmpty, !accepted.isEmpty else { return nil }
            return LoggedEvent(day: day, event: .suggestionAccepted(typed: typed, accepted: accepted))
        case "cr":
            guard fields.count == 5 else { return nil }
            let original = unescape(fields[3]), applied = unescape(fields[4])
            guard !original.isEmpty, !applied.isEmpty else { return nil }
            return LoggedEvent(day: day, event: .correctionReverted(original: original, applied: applied))
        case "wt":
            guard fields.count == 4 else { return nil }
            let word = unescape(fields[3])
            guard !word.isEmpty else { return nil }
            return LoggedEvent(day: day, event: .wordTapped(word: word))
        case "ts":
            guard fields.count == 6, let dx = Double(fields[4]), let dy = Double(fields[5]) else { return nil }
            let key = unescape(fields[3])
            guard key.count == 1, let keyChar = key.first else { return nil }
            return LoggedEvent(day: day, event: .touchSample(keyChar: keyChar, dx: dx, dy: dy))
        default:
            return nil  // unknown event code or schema version — skip
        }
    }

    static func escape(_ field: String) -> String {
        var out = ""
        out.reserveCapacity(field.count)
        for char in field {
            switch char {
            case "\\": out += "\\\\"
            case "\t": out += "\\t"
            case "\n": out += "\\n"
            case "\r": out += "\\r"
            default: out.append(char)
            }
        }
        return out
    }

    static func unescape(_ field: String) -> String {
        guard field.contains("\\") else { return field }
        var out = ""
        out.reserveCapacity(field.count)
        var iterator = field.makeIterator()
        while let char = iterator.next() {
            if char == "\\", let next = iterator.next() {
                switch next {
                case "\\": out.append("\\")
                case "t": out.append("\t")
                case "n": out.append("\n")
                case "r": out.append("\r")
                default:
                    out.append(char)
                    out.append(next)
                }
            } else {
                out.append(char)
            }
        }
        return out
    }

    // MARK: - Raw append

    /// One `open(O_RDWR|O_APPEND|O_CREAT)` + one `write(2)` of the whole
    /// batch. Writes the generation header first when the file is empty, and
    /// a healing `\n` first when the current tail is torn.
    private func appendRaw(_ lines: String) throws {
        let fd = url.withUnsafeFileSystemRepresentation { path -> Int32 in
            guard let path else { return -1 }
            return open(path, O_RDWR | O_APPEND | O_CREAT, 0o600)
        }
        guard fd >= 0 else {
            throw EventLogError.ioError("open failed: errno \(errno)")
        }
        defer { close(fd) }

        var status = stat()
        guard fstat(fd, &status) == 0 else {
            throw EventLogError.ioError("fstat failed: errno \(errno)")
        }

        var payload = ""
        if status.st_size == 0 {
            payload += Self.headerLine(generation: UUID())
        } else {
            var lastByte: UInt8 = 0
            let readCount = pread(fd, &lastByte, 1, status.st_size - 1)
            if readCount == 1 && lastByte != 0x0A {
                payload += "\n"  // heal a torn tail into an isolated garbage line
            }
        }
        payload += lines

        let data = Array(payload.utf8)
        var written = 0
        while written < data.count {
            let result = data[written...].withUnsafeBytes { buffer -> Int in
                write(fd, buffer.baseAddress, buffer.count)
            }
            if result < 0 {
                if errno == EINTR { continue }
                throw EventLogError.ioError("write failed: errno \(errno)")
            }
            written += result
        }
    }
}

File: AndroidClient/lib/engine/src/main/kotlin/is/solberg/lyklabord/engine/learning/EventLog.kt
package `is`.solberg.lyklabord.engine.learning

import java.io.File
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.nio.file.StandardOpenOption
import java.util.Locale
import java.util.UUID

sealed class EventLogError(message: String) : Exception(message) {
    class InvalidContent(detail: String) : EventLogError("Invalid event content: $detail")
    class IoError(detail: String) : EventLogError("Event log I/O error: $detail")
}

/** Append-only, crash-tolerant schema-v1 event log. */
class EventLog(
    val file: File,
    val dayProvider: () -> Int = { DayBucket.current() },
) {
    val url: File get() = file

    data class ConsumedMarker(val generation: UUID, val offset: Long) {
        companion object {
            val none: ConsumedMarker = ConsumedMarker(UUID(0L, 0L), 0L)
        }
    }

    data class ReadResult(val events: List<LoggedEvent>, val endMarker: ConsumedMarker, val skippedLines: Int)

    companion object {
        const val maxWordLength: Int = 64

        fun isLearnableWord(word: String): Boolean {
            val cps = word.codePoints().toArray()
            if (cps.isEmpty() || cps.size > maxWordLength) return false
            var hasLetter = false
            for (cp in cps) {
                if (Character.isWhitespace(cp) || Character.isISOControl(cp)) return false
                if (isEmoji(cp)) return false
                if (Character.isLetter(cp)) hasLetter = true
            }
            return hasLetter
        }

        private fun isEmoji(cp: Int): Boolean =
            cp in 0x1F000..0x1FAFF || cp in 0x2600..0x27BF || cp in 0x2300..0x23FF ||
                cp in 0x00A9..0x00AE || cp == 0x203C || cp == 0x2049 || cp == 0x2122 ||
                cp == 0x2139 || cp in 0x3030..0x303D || cp in 0x3297..0x3299

        internal fun headerLine(generation: UUID): String = "#gen\t$generation\n"

        internal fun parseHeader(bytes: ByteArray): Pair<UUID, Int?> {
            val prefix = "#gen\t".toByteArray(StandardCharsets.UTF_8)
            if (bytes.size < prefix.size || !bytes.copyOfRange(0, prefix.size).contentEquals(prefix)) {
                return ConsumedMarker.none.generation to 0
            }
            val newline = bytes.indexOf(0x0A.toByte())
            if (newline < 0) return ConsumedMarker.none.generation to null
            val uuidText = String(bytes, prefix.size, newline - prefix.size, StandardCharsets.UTF_8)
            val uuid = runCatching { UUID.fromString(uuidText) }.getOrNull() ?: ConsumedMarker.none.generation
            return uuid to (newline + 1)
        }
        private fun nextNewline(bytes: ByteArray, start: Int): Int {
            for (index in start until bytes.size) if (bytes[index] == 0x0A.toByte()) return index
            return -1
        }

        internal fun escape(field: String): String = buildString(field.length) {
            for (ch in field) when (ch) {
                '\\' -> append("\\\\")
                '\t' -> append("\\t")
                '\n' -> append("\\n")
                '\r' -> append("\\r")
                else -> append(ch)
            }
        }

        internal fun unescape(field: String): String {
            if (!field.contains('\\')) return field
            val out = StringBuilder(field.length)
            var i = 0
            while (i < field.length) {
                val ch = field[i]
                if (ch == '\\' && i + 1 < field.length) {
                    when (val next = field[i + 1]) {
                        '\\' -> out.append('\\')
                        't' -> out.append('\t')
                        'n' -> out.append('\n')
                        'r' -> out.append('\r')
                        else -> { out.append(ch); out.append(next) }
                    }
                    i += 2
                } else { out.append(ch); i++ }
            }
            return out.toString()
        }
    }

    fun append(event: LearningEvent) = append(contentsOf = listOf(event))

    fun append(contentsOf: List<LearningEvent>) {
        if (contentsOf.isEmpty()) return
        val day = dayProvider()
        val payload = buildString { contentsOf.forEach { append(encodeLine(it, day)) } }
        appendRaw(payload)
    }

    fun read(after: ConsumedMarker? = null): ReadResult {
        if (!file.exists()) return ReadResult(emptyList(), ConsumedMarker.none, 0)
        val bytes = try { file.readBytes() } catch (e: Exception) { throw EventLogError.IoError("read failed: $e") }
        val (generation, headerEnd) = parseHeader(bytes)
        val header = headerEnd ?: return ReadResult(emptyList(), ConsumedMarker.none, 0)
        var start = header
        if (after != null && after.generation == generation && after.offset >= header && after.offset <= bytes.size) {
            start = after.offset.toInt()
        }
        val events = mutableListOf<LoggedEvent>()
        var skipped = 0
        var cursor = start
        var consumedEnd = start
        while (cursor < bytes.size) {
            val newline = nextNewline(bytes, cursor)
            if (newline < 0) break
            val line = String(bytes, cursor, newline - cursor, StandardCharsets.UTF_8)
            val decoded = decodeLine(line)
            if (decoded != null) events += decoded else if (line.isNotEmpty()) skipped++
            cursor = newline + 1
            consumedEnd = cursor
        }
        return ReadResult(events, ConsumedMarker(generation, consumedEnd.toLong()), skipped)
    }

    fun truncate(consumedUpTo: ConsumedMarker): ConsumedMarker {
        if (!file.exists()) return consumedUpTo
        val bytes = try { file.readBytes() } catch (e: Exception) { throw EventLogError.IoError("read for truncate failed: $e") }
        val (generation, headerEnd) = parseHeader(bytes)
        if (headerEnd == null || generation != consumedUpTo.generation) return consumedUpTo
        val tailStart = consumedUpTo.offset.coerceIn(0L, bytes.size.toLong()).toInt()
        val newGeneration = UUID.randomUUID()
        val header = headerLine(newGeneration).toByteArray(StandardCharsets.UTF_8)
        val output = ByteArray(header.size + bytes.size - tailStart)
        header.copyInto(output)
        if (tailStart < bytes.size) bytes.copyInto(output, header.size, tailStart)
        atomicWrite(output)
        return ConsumedMarker(newGeneration, header.size.toLong())
    }

    private fun encodeLine(event: LearningEvent, day: Int): String {
        fun validated(word: String, field: String): String {
            if (!isLearnableWord(word)) throw EventLogError.InvalidContent("$field is not a learnable word")
            return escape(word)
        }
        val fields = when (event) {
            is LearningEvent.WordCommitted -> listOf("wc", validated(event.word, "word"),
                if (event.previousWord != null && isLearnableWord(event.previousWord)) escape(event.previousWord) else "",
                event.languageHint.rawValue)
            is LearningEvent.SuggestionAccepted -> listOf("sa", validated(event.typed, "typed"), validated(event.accepted, "accepted"))
            is LearningEvent.CorrectionReverted -> listOf("cr", validated(event.original, "original"), validated(event.applied, "applied"))
            is LearningEvent.WordTapped -> listOf("wt", validated(event.word, "word"))
            is LearningEvent.TouchSample -> listOf("ts", escape(event.keyChar.toString()),
                String.format(Locale.ROOT, "%.4f", event.dx), String.format(Locale.ROOT, "%.4f", event.dy))
        }
        return "1\t$day\t${fields.joinToString("\t")}\n"
    }

    private fun decodeLine(line: String): LoggedEvent? {
        val fields = line.split("\t", ignoreCase = false, limit = Int.MAX_VALUE)
        if (fields.size < 3 || fields[0] != "1") return null
        val day = fields[1].toIntOrNull() ?: return null
        return when (fields[2]) {
            "wc" -> if (fields.size == 6) {
                val word = unescape(fields[3]); val prev = fields[4].takeIf { it.isNotEmpty() }?.let(::unescape)
                val hint = when (fields[5]) { "is" -> LanguageHint.ICELANDIC; "en" -> LanguageHint.ENGLISH; "un" -> LanguageHint.UNKNOWN; else -> return null }
                if (word.isEmpty()) null else LoggedEvent(day, LearningEvent.WordCommitted(word, prev, hint))
            } else null
            "sa" -> if (fields.size == 5) {
                val typed = unescape(fields[3]); val accepted = unescape(fields[4])
                if (typed.isEmpty() || accepted.isEmpty()) null else LoggedEvent(day, LearningEvent.SuggestionAccepted(typed, accepted))
            } else null
            "cr" -> if (fields.size == 5) {
                val original = unescape(fields[3]); val applied = unescape(fields[4])
                if (original.isEmpty() || applied.isEmpty()) null else LoggedEvent(day, LearningEvent.CorrectionReverted(original, applied))
            } else null
            "wt" -> if (fields.size == 4) unescape(fields[3]).takeIf { it.isNotEmpty() }?.let { LoggedEvent(day, LearningEvent.WordTapped(it)) } else null
            "ts" -> if (fields.size == 6) {
                val key = unescape(fields[3]); val dx = fields[4].toDoubleOrNull(); val dy = fields[5].toDoubleOrNull()
                if (key.codePointCount(0, key.length) != 1 || dx == null || dy == null) null
                else LoggedEvent(day, LearningEvent.TouchSample(key[0], dx, dy))
            } else null
            else -> null
        }
    }

    private fun appendRaw(lines: String) {
        file.parentFile?.mkdirs()
        try {
            java.nio.channels.FileChannel.open(file.toPath(), StandardOpenOption.CREATE, StandardOpenOption.WRITE, StandardOpenOption.READ).use { channel ->
                val size = channel.size()
                val prefix = when {
                    size == 0L -> headerLine(UUID.randomUUID())
                    else -> {
                        val one = ByteBuffer.allocate(1)
                        channel.position(size - 1L); channel.read(one); one.flip()
                        if (one.get().toInt() != 0x0A) "\n" else ""
                    }
                }
                val data = (prefix + lines).toByteArray(StandardCharsets.UTF_8)
                channel.position(channel.size())
                var written = 0
                while (written < data.size) written += channel.write(ByteBuffer.wrap(data, written, data.size - written))
            }
        } catch (e: EventLogError) { throw e } catch (e: Exception) { throw EventLogError.IoError("append failed: $e") }
    }

    private fun atomicWrite(bytes: ByteArray) {
        file.parentFile?.mkdirs()
        val tmp = File(file.parentFile ?: File("."), ".${file.name}.${UUID.randomUUID()}.tmp")
        try {
            Files.write(tmp.toPath(), bytes, StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE)
            try { Files.move(tmp.toPath(), file.toPath(), StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING) }
            catch (_: Exception) { Files.move(tmp.toPath(), file.toPath(), StandardCopyOption.REPLACE_EXISTING) }
        } catch (e: Exception) { tmp.delete(); throw EventLogError.IoError("truncate rewrite failed: $e") }
    }
}
